pkg/processor/sourceprocessor/source_category_filler.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2021 Sumo Logic, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sourceprocessor
+
+import (
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/collector/model/pdata"
+)
+
+type sourceCategoryFiller struct {
+	attributeName      string
+	valueTemplate      string
+	templateAttributes []string
+	prefix             string
+	dashReplacement    string
+	annotationPrefix   string
+}
+
+func newSourceCategoryFiller(
+	attributeName string,
+	cfg *Config,
+) sourceCategoryFiller {
+
+	valueTemplate := cfg.SourceCategory
+	templateAttributes := extractTemplateAttributes(valueTemplate)
+
+	return sourceCategoryFiller{
+		attributeName:      attributeName,
+		valueTemplate:      valueTemplate,
+		templateAttributes: templateAttributes,
+		prefix:             cfg.SourceCategoryPrefix,
+		dashReplacement:    cfg.SourceCategoryReplaceDash,
+		annotationPrefix:   cfg.AnnotationPrefix,
+	}
+}
+
+func extractTemplateAttributes(template string) []string {
+	attributes := make([]string, 0)
+	attributeMatches := formatRegex.FindAllStringSubmatch(template, -1)
+	for _, matchset := range attributeMatches {
+		attributes = append(attributes, matchset[1])
+	}
+	return attributes
+}
+
+func (f *sourceCategoryFiller) fill(attributes *pdata.AttributeMap) {
```

Comment:
comment? Can we describe how it works?

Reply:
How about this? https://github.com/SumoLogic/sumologic-otel-collector/pull/257/commits/caa1ad024eb079198d896d7a870383f8e19aff07

Reply:
Cool, can we specify that this is actually the order of precedence?

Reply:
Anything else I could add? https://github.com/SumoLogic/sumologic-otel-collector/pull/257/commits/13a9d1d05b332300482ddbd84b539e9fd9ce524b